File: apps/server/src/line/handlers.ts
import type { Message, MessageEvent, WebhookEvent } from '@line/bot-sdk';

import { listSweets } from '../services/sweetService.js';
import { buildCustomerServiceMessage, buildDefaultMessage, buildRulesMessage, buildSweetCarousel } from './messages.js';
import { lineClient } from './client.js';

const keywordMap: Record<string, () => Promise<Message | Message[]>> = {
  甜心列表: async () => {
    const sweets = await listSweets();
    if (!sweets.length) {
      return {
        type: 'text',
        text: '目前沒有上架的甜心，請稍後再試或聯絡客服唷。',
      };
    }
    return buildSweetCarousel(sweets);
  },
  預約規則: async () => buildRulesMessage(),
  客服: async () => buildCustomerServiceMessage(),
};

function isTextMessageEvent(event: WebhookEvent): event is MessageEvent & { message: { type: 'text'; text: string } } {
  return event.type === 'message' && event.message.type === 'text';
}

export async function handleLineEvent(event: WebhookEvent) {
  if (!isTextMessageEvent(event)) {
    return;
  }

  const text = event.message.text.trim();
  const builder = keywordMap[text];

  let reply: Message | Message[] = buildDefaultMessage();

  if (builder) {
    reply = await builder();
  }

  if (Array.isArray(reply)) {
    await lineClient.replyMessage(event.replyToken, reply);
  } else {
    await lineClient.replyMessage(event.replyToken, [reply]);
  }
}

File: apps/server/linebot/handlers.py
from __future__ import annotations

import logging
from typing import Any, Dict

from django.conf import settings

from api import services
from api.models import Location, Sweet
from .client import client
from . import messages

logger = logging.getLogger(__name__)


def handle_event(event: Dict[str, Any]) -> None:
    if event.get("type") != "message":
        return
    message = event.get("message") or {}
    if message.get("type") != "text":
        return

    text = (message.get("text") or "").strip()

    reply_message = messages.build_default_message()
    if text == "甜心列表":
        locations = list(Location.objects.order_by("name"))
        reply_message = messages.build_location_menu(locations)
    elif location := find_location_from_text(text):
        sweets = list(services.list_sweets(location_slug=location.slug))
        reply_message = messages.build_sweet_carousel(sweets, title=f"{location.name}甜心")
    elif text == "預約規則":
        reply_message = messages.build_rules_message()
    elif text == "客服":
        reply_message = messages.build_customer_service_message()

    try:
        client.reply_message(event["replyToken"], [reply_message])
    except Exception as exc:  # pragma: no cover - network failure already logged
        logger.error("Failed to send LINE reply: %s", exc)


def find_location_from_text(text: str) -> Location | None:
    normalized = text.replace("甜心列表", "").replace("甜心", "").replace("地區", "").strip()
    if not normalized:
        normalized = text

    lowered = normalized.lower()
    for location in Location.objects.all():
        if location.name in text:
            return location
        if location.slug and location.slug.lower() in lowered:
            return location
    return None
